packages/components/src/date-time/time.js

```diff
@@ -23,347 +28,258 @@ import TimeZone from './timezone';
-		this.changeDate( newDate );
-	}
-
-	updateDay() {
-		const { day, date } = this.state;
-		const value = parseInt( day, 10 );
-		if ( value === date.date() ) {
-			return;
-		}
-		if ( ! isInteger( value ) || value < 1 || value > 31 ) {
-			this.syncState( this.props );
-			return;
-		}
-		const newDate = date.clone().date( value );
-		this.changeDate( newDate );
-	}
-
-	updateMonth() {
-		const { month, date } = this.state;
-		const value = parseInt( month, 10 );
-		if ( value === date.month() + 1 ) {
-			return;
-		}
-		if ( ! isInteger( value ) || value < 1 || value > 12 ) {
-			this.syncState( this.props );
-			return;
-		}
-		const newDate = date.clone().month( value - 1 );
-		this.changeDate( newDate );
-	}
-
-	updateYear() {
-		const { year, date } = this.state;
-		const value = parseInt( year, 10 );
-		if ( value === date.year() ) {
-			return;
-		}
-		if ( ! isInteger( value ) || value < 0 || value > 9999 ) {
-			this.syncState( this.props );
-			return;
-		}
-		const newDate = date.clone().year( value );
-		this.changeDate( newDate );
-	}
-
-	updateAmPm( value ) {
+	function updateAmPm( value ) {
 		return () => {
-			const { am, date, hours } = this.state;
 			if ( am === value ) {
 				return;
 			}
-			let newDate;
-			if ( value === 'PM' ) {
-				newDate = date
-					.clone()
-					.hours( ( ( parseInt( hours, 10 ) % 12 ) + 12 ) % 24 );
-			} else {
-				newDate = date.clone().hours( parseInt( hours, 10 ) % 12 );
-			}
-			this.changeDate( newDate );
-		};
-	}
 
-	onChangeDay( event ) {
-		this.setState( { day: event.target.value } );
-	}
+			const parsedHours = parseInt( hours, 10 );
 
-	onChangeMonth( event ) {
-		this.setState( { month: event.target.value } );
-	}
+			const newDate = date
+				.clone()
+				.hours(
+					value === 'PM'
+						? ( ( parsedHours % 12 ) + 12 ) % 24
+						: parsedHours % 12
+				);
 
-	onChangeYear( event ) {
-		this.setState( { year: event.target.value } );
-	}
-
-	onChangeHours( event ) {
-		this.setState( { hours: event.target.value } );
+			changeDate( newDate );
+		};
 	}
 
-	onChangeMinutes( event ) {
-		const minutes = event.target.value;
-		this.setState( {
-			minutes: minutes === '' ? '' : ( '0' + minutes ).slice( -2 ),
-		} );
-	}
+	const dayFormat = (
+		<div className="components-datetime__time-field components-datetime__time-field-day">
+			<NumberInput
+				aria-label={ __( 'Day' ) }
+				className="components-datetime__time-field-day-input"
+				type="number"
+				// The correct function to call in moment.js is "date" not "day"
+				name="date"
+				value={ day }
+				step={ 1 }
+				min={ 1 }
+				max={ 31 }
+				onUpdate={ update }
+			/>
+		</div>
+	);
 
-	renderMonth( month ) {
-		return (
-			<div
-				key="render-month"
-				className="components-datetime__time-field components-datetime__time-field-month"
+	const monthFormat = (
+		<div className="components-datetime__time-field components-datetime__time-field-month">
+			<NumberInput
+				as="select"
```

Comment:
Maybe it's just the name, but it feels a little confusing to me to have a component called 'NumberInput' that renders a `<select>`.

I think renaming it to something that represents what it does would be good. Not 100% sure I know what it does ... validation? Maybe something like `ValidatedFormField`

Reply:
Yeah I was skeptical about this too. It does transform the input value to type of `number` though. I'll see what I can do!